### PR TITLE
[CTRLRFC-1194]Add hability to manage Kind and Polarity output trigger options to Hamamatsu camera

### PR DESCRIPTION
**Pull request by goldensgui**
Add getter and setter methods to allow management of the KIND and POLARITY options related to equipment output triggers :
 
- Getters:
    - enum Camera::Output_Trigger_Kind getOutputTriggerKind(int channel)
    - enum Camera::Output_Trigger_Polarity getOutputTriggerPolarity(int channel)

- Setters
    - void setOutputTriggerKind    (int channel, enum Output_Trigger_Kind in_output_trig_kind);                 
    - void setOutputTriggerPolarity(int in_channel,  enum Camera::Output_Trigger_Polarity in_output_trig_polarity);

- Enum created :
    - Output_Trigger_Kind
    - Output_Trigger_Polarity

### CODE REVIEW

File: include/HamamatsuCamera.h [hunk 1]
```diff
@@ -209,42 +229,54 @@ namespace lima
         void setViewMode(bool in_view_mode_activated,  ///< [in] view mode activation or not
                          int  in_views_number       ); ///< [in] number of views if view mode activated
 
+        //-- Output Triggers  control
+        void setOutputTriggerKind    (int channel,                                                   ///< [in] channel to set
+                                      enum Output_Trigger_Kind in_output_trig_kind);                 ///< [in] kind of the channel to set
```
Comment (soleil-ica) [inline include/HamamatsuCamera.h]: enum Output_Trigger_Kind  : missing **Camera::**

File: src/HamamatsuCamera.cpp [hunk 2]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+//=============================================================================
+//-----------------------------------------------------------------------------
+/// Return the output trigger kind of given channel by the current detector
+//-----------------------------------------------------------------------------
+enum Camera::Output_Trigger_Kind Camera::getOutputTriggerKind(int channel)
+{
+    DEB_MEMBER_FUNCT();
+
+    DEB_TRACE() << " Camera::Output_Trigger_Kind Camera::getOutputTriggerKind(int channel) : ..."; 
+    enum Output_Trigger_Kind result = Output_Trigger_Kind::Output_Trigger_Kind_Not_Supported;
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: Is **result** used somewhere ?

File: src/HamamatsuCamera.cpp [hunk 3]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    DCAMERR err;
+    double kindArraySize = 0;
+    enum Output_Trigger_Kind kind = Camera::Output_Trigger_Kind_Not_Supported;
+
+	// get property attribute that contains the Kind (that may be an array)
+	DCAMPROP_ATTR	basepropattr;
+	memset(&basepropattr, 0, sizeof(basepropattr));
+	basepropattr.cbSize = sizeof(basepropattr);
+	basepropattr.iProp = DCAM_IDPROP_OUTPUTTRIGGER_KIND;
+	err = dcamprop_getattr(m_camera_handle, &basepropattr);
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: 1/ could this code be replaced by the getPropertyData function ?
2/ fix the formatting: ie start is not aligned

File: src/HamamatsuCamera.cpp [hunk 4]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    DCAMERR err;
+    double polarityArraySize = 0;
+    enum Output_Trigger_Polarity polarity = Camera::Output_Trigger_Polarity_Not_Supported;
+
+	// get property attribute that contains the Polarity (that may be an array)
+	DCAMPROP_ATTR	basepropattr;
+	memset(&basepropattr, 0, sizeof(basepropattr));
+	basepropattr.cbSize = sizeof(basepropattr);
+	basepropattr.iProp = DCAM_IDPROP_OUTPUTTRIGGER_POLARITY;
+	err = dcamprop_getattr(m_camera_handle, &basepropattr);
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: 1/ could this code be replaced by the getPropertyData function ?
2/ fix the formatting: ie start is not aligned

File: src/HamamatsuCamera.cpp [hunk 5]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+                
+                
+                int32 value = static_cast<int32>(tmp);
+
+                switch(value) 
+                {
+                    case DCAMPROP_OUTPUTTRIGGER_POLARITY__NEGATIVE     : polarity = Camera::Output_Trigger_Polarity_Negative ; break;
+                    case DCAMPROP_OUTPUTTRIGGER_POLARITY__POSITIVE     : polarity = Camera::Output_Trigger_Polarity_Positive ; break;
+                    default: break; // result will be Output_Trigger_Polarity_Not_Supported
+
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: Fix alignement of code + empty lines uneeded

File: src/HamamatsuCamera.cpp [hunk 6]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+void Camera::getPropertyData(int32 property, int32 & array_base, int32 & step_element){
+    
+    DCAMERR  err ;
+	// get property attribute
+	DCAMPROP_ATTR	basepropattr;
+	memset(&basepropattr, 0, sizeof(basepropattr));
+	basepropattr.cbSize = sizeof(basepropattr);
+	basepropattr.iProp = property;
+	err = dcamprop_getattr(m_camera_handle, &basepropattr);
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: idem: alignement .

File: src/HamamatsuCamera.cpp [hunk 7]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    if (!failed(err))
+	{
+        array_base = basepropattr.iProp_ArrayBase;
+        step_element = basepropattr.iPropStep_Element;
+
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: empty line to remove

File: src/HamamatsuCamera.cpp [hunk 8]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    err = dcamprop_setvalue( m_camera_handle,property_id , static_cast<double>(kind) );
+
+    if( failed(err) )
+    {
+        if((err == DCAMERR_INVALIDPROPERTYID)||(err == DCAMERR_NOTSUPPORT))
+        {
+            manage_trace( deb, "Unable to set the Output trigger Kind", err, 
+                               "dcamprop_setvalue", "DCAM_IDPROP_OUTPUTTRIGGER_KIND[%d] %d", channel, kind);
+        }
+
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: remove empty line

File: src/HamamatsuCamera.cpp [hunk 9]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    // set the kind
+    err = dcamprop_setvalue( m_camera_handle,property_id , static_cast<double>(kind) );
+
+    if( failed(err) )
+    {
+        if((err == DCAMERR_INVALIDPROPERTYID)||(err == DCAMERR_NOTSUPPORT))
+        {
+            manage_trace( deb, "Unable to set the Output trigger Kind", err, 
+                               "dcamprop_setvalue", "DCAM_IDPROP_OUTPUTTRIGGER_KIND[%d] %d", channel, kind);
+        }
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp:3191]: Why not a THROW_HW_ERROR here while there is one under ?

File: src/HamamatsuCamera.cpp [hunk 10]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+    // {
+    //     case Output_Trigger_Polarity_Negative : polarity = DCAMPROP_OUTPUTTRIGGER_POLARITY__NEGATIVE; break;
+    //     case Output_Trigger_Polarity_Positive : polarity = DCAMPROP_OUTPUTTRIGGER_POLARITY__POSITIVE; break;
+
+    //     default : 
+    //     manage_error( deb, "Unable to set the Output trigger Polarity", DCAMERR_NONE, 
+    //                        "", "in_output_trig_polarity is unknown %d", static_cast<int>(in_output_trig_polarity));
+    //     THROW_HW_ERROR(Error) << "Unable to set the Output trigger Polarity";
+    //     break;
+    // }
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp]: to be removed

File: src/HamamatsuCamera.cpp [hunk 11]
```diff
@@ -2977,4 +2977,290 @@ void Camera::setHighDynamicRangeEnabled(const bool & in_enabled)
...
+
+    if (Output_Trigger_Polarity_Negative == in_output_trig_polarity)
+    {
+        polarity = DCAMPROP_OUTPUTTRIGGER_POLARITY__NEGATIVE;
+    }
+
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp:3214]: remove blank lines here

File: include/HamamatsuCamera.h [hunk 12]
```diff
@@ -209,42 +229,54 @@ namespace lima
...
         short int getReadoutSpeed(void) const;
 		void      setReadoutSpeed(const short int readout_speed); ///< [in]  new readout speed
-
-        std::string getReadoutSpeedLabelFromValue(const short int in_readout_speed) const;
-        short int   getReadoutSpeedFromLabel     (const std::string & in_readout_speed_label) const;
+
+        std::string getReadoutSpeedLabelFromValue(const short int in_readout_speed) const;
+        short int   getReadoutSpeedFromLabel     (const std::string & in_readout_speed_label) const;
+
+        void getPropertyData(int32 property, int32 & array_base, int32 & step_element);
```
Comment (soleil-ica) [inline include/HamamatsuCamera.h:286]: add a comment for this function. with param description

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by soleil-ica** [comment]
Generaly: fix alignement of start of code and remove useless blank lines

File: src/HamamatsuCamera.cpp [hunk 13]
```diff
@@ -3052,78 +3051,80 @@ enum Camera::Output_Trigger_Kind Camera::getOutputTriggerKind(int channel)
...
-	memset(&basepropattr, 0, sizeof(basepropattr));
-	basepropattr.cbSize = sizeof(basepropattr);
-	basepropattr.iProp = DCAM_IDPROP_OUTPUTTRIGGER_POLARITY;
-	err = dcamprop_getattr(m_camera_handle, &basepropattr);
+    // get property attribute that contains the Polarity (that may be an array)
+    DCAMPROP_ATTR basepropattr;
+    memset(&basepropattr, 0, sizeof(basepropattr));
+    basepropattr.cbSize = sizeof(basepropattr);
+    basepropattr.iProp = DCAM_IDPROP_OUTPUTTRIGGER_POLARITY;
+    err = dcamprop_getattr(m_camera_handle, &basepropattr);
```
Comment (soleil-ica) [inline src/HamamatsuCamera.cpp:3069]: possibility to use getPropertyData() ?

**Review by ghost — APPROVED** [review]

File: src/HamamatsuCamera.cpp (hunk 13 above)
Comment (goldensgui) [reply src/HamamatsuCamera.cpp:3069]: Yes, but in this case we manage error from get Hamamatsu data for kind and polarity

**Review by ghost — APPROVED** [review]

File: src/HamamatsuCamera.cpp (hunk 13 above)
Comment (goldensgui) [reply src/HamamatsuCamera.cpp:3069]: getPropertyData is not yet able to get the number of element when the requested attribute is an array. Dealing with the attribute structure directly allow us to get that number of element which is then used to ensure that the input value (channel) is not beyond the sensor capabilities.

Such behaviour may be added if required.

**Review by ghost — APPROVED** [review]

**Pull request merged**